src/common/SharedMemory.cpp

```diff
@@ -0,0 +1,230 @@
+		m_mapping = mmap(nullptr, m_size, mappingProtection, MAP_SHARED, fd.get(), 0);
+		if (m_mapping == MAP_FAILED)
+		{
+			throw std::system_error{errno, std::generic_category(), "SharedMemoryImpl: mmap() failed"};
+		}
+	}
+
+	SharedMemoryImpl(const std::string& key, std::size_t size, bool readOnly) :
+		m_key{"/" + key},
+		m_size{size}
+	{
+		const auto fd = FileDescriptor{
+			retryWhileInterrupted([&]() noexcept { return shm_open(m_key.c_str(), O_RDWR | O_CREAT | O_EXCL, 0600); })
+		};
+		if (fd.get() == -1)
+		{
+			throw std::system_error{errno, std::generic_category(), "SharedMemoryImpl: shm_open() failed"};
+		}
+		m_object.reset(m_key.c_str());
+		if (retryWhileInterrupted([&]() noexcept { return ftruncate(fd.get(), m_size); }) == -1)
+		{
+			throw std::system_error{errno, std::generic_category(), "SharedMemoryImpl: ftruncate() failed"};
+		}
+		const auto mappingProtection = readOnly ? PROT_READ : PROT_READ | PROT_WRITE;
+		m_mapping = mmap(nullptr, m_size, mappingProtection, MAP_SHARED, fd.get(), 0);
+		if (m_mapping == MAP_FAILED)
+		{
+			throw std::system_error{errno, std::generic_category(), "SharedMemoryImpl: mmap() failed"};
+		}
+	}
+
+	SharedMemoryImpl(const SharedMemoryImpl&) = delete;
+	SharedMemoryImpl& operator=(const SharedMemoryImpl&) = delete;
+
+	~SharedMemoryImpl()
+	{
+		if (m_mapping) { munmap(m_mapping, m_size); }
+	}
+
+	void* get() { return m_mapping; }
+
+private:
+	std::string m_key;
+	std::size_t m_size;
+	void* m_mapping;
+	ShmObject m_object;
+};
+
+#else
+
+class SharedMemoryImpl
+{
+public:
+	SharedMemoryImpl(const std::string& key, bool readOnly) :
+		m_shm{QString::fromStdString(key)}
+	{
+		const auto mode = readOnly ? QSharedMemory::ReadOnly : QSharedMemory::ReadWrite;
+		if (!m_shm.attach(mode))
+		{
+			throw std::runtime_error{"SharedMemoryImpl: QSharedMemory::attach() failed"};
+		}
+	}
+
+	SharedMemoryImpl(const std::string& key, std::size_t size, bool readOnly) :
+		m_shm{QString::fromStdString(key)}
+	{
+		const auto mode = readOnly ? QSharedMemory::ReadOnly : QSharedMemory::ReadWrite;
+		if (!m_shm.create(size, mode))
+		{
+			throw std::runtime_error{"SharedMemoryImpl: QSharedMemory::create() failed"};
+		}
+	}
+
+	SharedMemoryImpl(const SharedMemoryImpl&) = delete;
+	SharedMemoryImpl& operator=(const SharedMemoryImpl&) = delete;
+
+	void* get() { return m_shm.data(); }
+
+private:
+	QSharedMemory m_shm;
+};
+
+#endif
+
+SharedMemoryData::SharedMemoryData() noexcept
+{ }
+
+SharedMemoryData::SharedMemoryData(std::string&& key, bool readOnly) :
+	m_key{std::move(key)},
+	m_impl{std::make_unique<SharedMemoryImpl>(m_key, readOnly)},
+	m_ptr{m_impl->get()}
+{ }
+
+SharedMemoryData::SharedMemoryData(std::string&& key, std::size_t size, bool readOnly) :
+	m_key{std::move(key)},
+	m_impl{std::make_unique<SharedMemoryImpl>(m_key, std::max(size, std::size_t{1}), readOnly)},
+	m_ptr{m_impl->get()}
+{ }
+
+SharedMemoryData::~SharedMemoryData() { }
```

Comment:
I see no use in defining a empty deconstructor. I think it is best to let it be implicity defined, which is of course empty.
```suggestion
```